File: ruby/easy/min_stack.rb
class MinStack

=begin
    initialize your data structure here.
=end
    def initialize()
        @stack = Array.new
    end


=begin
    :type x: Integer
    :rtype: Void
=end
    def push(x)
        if @stack.empty?
            @stack.push([x, x])
        else
            @stack.push([x, [x, @stack.last[1]].min])
        end
    end


=begin
    :rtype: Void
=end
    def pop()
        @stack.pop unless @stack.empty?
    end


=begin
    :rtype: Integer
=end
    def top()
        @stack.last[0] unless @stack.empty?
    end


=begin
    :rtype: Integer
=end
    def get_min()
        @stack.last[1] unless @stack.empty?
    end


end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()

File: javascript/easy/minStack.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.stack = [];
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    if(this.stack.length == 0) {
        this.stack.push([x, x]);
    } else {
        let top = this.stack[this.stack.length-1];
        this.stack.push([x, Math.min(x, top[1])]);
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if(this.stack.length > 0) {
        this.stack.pop();
    }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    if(this.stack.length > 0) {
        return this.stack[this.stack.length-1][0];
    }
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    if(this.stack.length > 0) {
        return this.stack[this.stack.length-1][1];
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = Object.create(MinStack).createNew()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
